pkg/bgpv1/agent/controller.go

```diff
@@ -0,0 +1,364 @@
+	// and provides a declarative API for configuring BGP peers.
+	BGPMgr BGPRouterManager
+}
+
+// ControllerOpt is a signature for defining configurable options for a
+// Controller
+type ControllerOpt func(*Controller)
+
+// NewController constructs a new BGP Control Plane Controller.
+//
+// When the constructor returns the Controller will be actively watching for
+// events and configuring BGP related sub-systems.
+//
+// The constructor requires an implementation of BGPRouterManager to be provided.
+// This implementation defines which BGP backend will be used (GoBGP, FRR, Bird, etc...)
+// NOTE: only GoBGP currently implemented.
+//
+// Cancel the provided CTX to stop the Controller.
+func NewController(ctx context.Context, rtMgr BGPRouterManager, opts ...ControllerOpt) (*Controller, error) {
+	// create the signaler we will use
+	sig := NewSignaler()
+
+	// informers setup
+	clientset := k8s.CiliumClient()
+	factory := externalversions.NewSharedInformerFactory(clientset, defaultInformerResync)
+	k8sfactory := informers.NewSharedInformerFactory(k8s.Client(), defaultInformerResync)
+
+	nodeLister := k8sfactory.Core().V1().Nodes().Lister()
+	nodeInformer := k8sfactory.Core().V1().Nodes().Informer()
+	nodeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    sig.Event,
+		UpdateFunc: func(_ interface{}, _ interface{}) { sig.Event(struct{}{}) },
+		DeleteFunc: sig.Event,
+	})
+
+	policyLister := factory.Cilium().V2alpha1().CiliumBGPPeeringPolicies().Lister()
+	policyInformer := factory.Cilium().V2alpha1().CiliumBGPPeeringPolicies().Informer()
+	policyInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    sig.Event,
+		UpdateFunc: func(_ interface{}, _ interface{}) { sig.Event(struct{}{}) },
+		DeleteFunc: sig.Event,
+	})
+
+	c := Controller{
+		Sig:               sig,
+		NodeInformer:      nodeInformer,
+		NodeLister:        nodeLister,
+		BGPPolicyInformer: policyInformer,
+		PolicyLister:      policyLister,
+		BGPMgr:            rtMgr,
+	}
+
+	// apply any options.
+	for _, opt := range opts {
+		opt(&c)
+	}
+
+	go c.Run(ctx)
+	return &c, nil
+}
+
+// Run places the Controller into its control loop.
+//
+// Kubernetes shared informers are started just before entering the long running
+// loop.
+//
+// When new events trigger a signal the control loop will be evaluated.
+//
+// A cancel of the provided ctx will kill the control loop along with the running
+// informers.
+func (c *Controller) Run(ctx context.Context) {
+	var (
+		l = log.WithFields(logrus.Fields{
+			"component": "Controller.Run",
+		})
+	)
+	l.Debug("Starting informers")
+
+	// add an initial signal to kick things off
+	c.Sig.Event(struct{}{})
+
+	stop := make(chan struct{})
+	go c.NodeInformer.Run(stop)
+	go c.BGPPolicyInformer.Run(stop)
+
+	l.Info("Cilium BGP Control Plane Controller now running...")
+	for {
+		select {
+		case <-ctx.Done():
+			killCTX, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
```

Comment:
Any reason for this 1min?

Reply:
Nothing scientific. Just seems like a time frame that's long enough to end any concurrent processes, but short enough that a user waiting for its death doesn't walk away for coffee. 

Any suggestions here?

Reply:
I'd suggest plumbing this up to a (hidden) flag, so that if we suspect this is a problem, we can tweak it fairly easily rather than having to roll a whole new build for it. This can also help with testing, where we could manually set the values low to trigger & observe failures more quickly.

Also, shouldn't this context be created from the parent function's `ctx` rather than `context.Background()`? This way, when the daemon terminates, it will initiate the termination of this logic as well.

Reply:
@joestringer I'm hesitant to add a flag to Cilium which we just won't use, in an attempt to be conservative with the large flag list we have currently.

Since this directive will be primarily for development how about we slip in an env variable check for this? So that we just define an environment variable in the deployment.yaml when Cilium is deployed to override the value?